### PR TITLE
add a wrapper for CEVAE

### PR DESCRIPTION
**Pull request by ppstacy**
## Proposed changes

Add a wrapper for CEVAE method

## Types of changes

What types of changes does your code introduce to **CausalML**?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/uber/causalml/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/uber/causalml)
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc. This PR template is adopted from [appium](https://github.com/appium/appium).

### CODE REVIEW

**Comment by jeongyoonlee** [comment]
Could you add torch and pyro to the requirements.txt?

**Comment by ppstacy** [comment]
> Could you add torch and pyro to the requirements.txt?
Thanks Jeong! I updated it.

**Comment by ppstacy** [comment]
@fritzo Hi Fritz, hope you are doing well and happy new year!!

In this pull request, I added a wrapper for CEVAE using the implementation the pyro team had.  I also added an example notebook [here](https://github.com/uber/causalml/blob/b17a132eaadac55c5325aa9b41a77991f5db498c/examples/cevae_example.ipynb). For the IHDP dataset I followed the paper to replicate it 1000 times to run, somehow the results I got seems not very good. So I also tried on the [synthetic dataset](https://github.com/uber/causalml/pull/133) that you added into CausalML earlier and similar set up for the parameters as used in pyro, but I still got the results no better than random. I played with different options but still couldn't get it to work. I am wondering if you can help take a look at if I am using the right input parameters or anything else here is wrong.

For IHDP I used:
`outcome_dist = "exponential"
latent_dim = 20
hidden_dim = 20
num_epochs = 50
batch_size = 1000
learning_rate = 1e-3
learning_rate_decay = 0.1`

For synthetic dataset I used:
`outcome_dist = "normal"
latent_dim = 20
hidden_dim = 200
num_epochs = 50
batch_size = 100
learning_rate = 1e-3
learning_rate_decay = 0.1`

**Review by paullo0106 — COMMENTED** [review]
A minor suggestion for cell 90 in the notebook, do we want to update the logging level so it excludes DEBUG log (or fit with a much smaller number of epochs) and will be much more readable?

**Comment by ppstacy** [comment]
> A minor suggestion for cell 90 in the notebook, do we want to update the logging level so it excludes DEBUG log (or fit with a much smaller number of epochs) and will be much more readable overall?

Good suggestion Paul, I will update it to not showing so many logs info. Thanks!

File: causalml/inference/nn/cevae.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+import torch
+from pyro.contrib.cevae import CEVAE
+from causalml.inference.meta.utils import convert_pd_to_np
+
+logging.getLogger("pyro").setLevel(logging.DEBUG)
+logging.getLogger("pyro").handlers[0].setLevel(logging.DEBUG)
+
+
+class CEVAE_MODEL(object):
+    def __init__(self, outcome_dist="studentt", latent_dim=20, hidden_dim=20, num_epochs=50, num_layers=3,
```
Comment (fritzo) [inline causalml/inference/nn/cevae.py:35]: Note the default in Pyro is `hidden_dim=200`.

**Review by fritzo — COMMENTED** [review]
Hi @ppstacy, indeed it looks like the `CEVAE_MODEL` is not learning anything, since the loss does not decrease during `.fit()`. I'm not sure what's wrong (and it's been a while), but you might try:
- decreasing `num_layers` to 1 or 2
- increasing `hidden_dim` to 50, 100, or 200
- trying other distributions, say "normal", maybe with a log transform on the outcome?

File: causalml/inference/nn/cevae.py (hunk 1 above)
Comment (ppstacy) [reply causalml/inference/nn/cevae.py]: Thanks @fritzo! I updated the hidden_dim to 200.

**Comment by ppstacy** [comment]
I refreshed the notebook after playing with the hyperparameters. Now the model at least starts learning, but the performance is not very competitive vs. meta-learners. Maybe we can let the users further tune the hyperparameters to get better results.

File: causalml/inference/nn/cevae.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+[1] C. Louizos, U. Shalit, J. Mooij, D. Sontag, R. Zemel, M. Welling (2017).
+    | Causal Effect Inference with Deep Latent-Variable Models.
+    | http://papers.nips.cc/paper/7223-causal-effect-inference-with-deep-latent-variable-models.pdf
+    | https://github.com/AMLab-Amsterdam/CEVAE
+"""
+
+import logging
+import torch
+from pyro.contrib.cevae import CEVAE
```
Comment (jeongyoonlee) [inline causalml/inference/nn/cevae.py:27]: Let's import `CEVAE` as `CEVAEModel` and rename `CEVAE_MODDEL` class as `CEVAE`.

Comment (ppstacy) [reply causalml/inference/nn/cevae.py]: Thanks! Updated now

File: causalml/inference/nn/__init__.py [hunk 3]
```diff
@@ -1 +1,2 @@
-from .dragonnet import DragonNet
\ No newline at end of file
+from .dragonnet import DragonNet
+from .cevae import CEVAE_MODEL
```
Comment (jeongyoonlee) [inline causalml/inference/nn/__init__.py:2]: This should be changed to `CEVAE` to reflect the name change in `cevae.py`.

File: tests/test_cevae.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import pandas as pd
+import torch
+
+from causalml.inference.nn import CEVAE_MODEL
```
Comment (jeongyoonlee) [inline tests/test_cevae.py:4]: This should be changed to `CEVAE` to reflect the change in `cevae.py`.

File: causalml/inference/nn/__init__.py (hunk 3 above)
Comment (ppstacy) [reply causalml/inference/nn/__init__.py:2]: Thanks for catching it!

**Review by jeongyoonlee — COMMENTED** [review]
Please update `CEVAE_MODEL` in the example notebook as well.

**Comment by ppstacy** [comment]
> Please update `CEVAE_MODEL` in the example notebook as well.

Done, apologies for all the back and force. I should change them all at once. Thanks!

**Review by jeongyoonlee — APPROVED** [review]
LGTM. I will update code to inherit `BaseLearner` once merged. Thanks.
